src/stream/src/executor/source/fetch_executor.rs

```diff
@@ -0,0 +1,467 @@
+                                                .into(),
+                                        )
+                                    });
+                                    state_cache.extend(file_assignment);
+
+                                    // When both of state cache and state table are empty,
+                                    // the right arm of stream is a pending stream,
+                                    // and is_datastream_empty is set to true,
+                                    // and a new reader should be built.
+                                    if is_datastream_empty {
+                                        Self::try_replace_with_new_reader(
+                                            &mut is_datastream_empty,
+                                            &state_store_handler,
+                                            &mut state_cache,
+                                            core.column_ids.clone(),
+                                            self.build_source_ctx(&source_desc, core.source_id),
+                                            &source_desc,
+                                            &mut stream,
+                                            &mut state_table_iter,
+                                        )
+                                        .await?;
+                                    }
+                                }
+                                _ => unreachable!(),
+                            }
+                        }
+                        // StreamChunk from FsSourceReader, and the reader reads only one file.
+                        // If the file read out, replace with a new file reader.
+                        Either::Right(StreamChunkWithState {
+                            chunk,
+                            split_offset_mapping,
+                        }) => {
+                            let mapping = split_offset_mapping.unwrap();
+                            debug_assert_eq!(mapping.len(), 1);
+
+                            let (split_id, offset) = mapping.iter().nth(0).unwrap();
+                            if !state_cache.contains_key(split_id) {
+                                // The data chunk may haven't been produced by reader yet when
+                                // a barrier arrives and the state cache is flushed and cleared.
+                                // Here we expect the fs_split lies in the state store and
+                                // store it into the state cache.
+                                let row = state_store_handler
+                                    .get(split_id.to_owned())
+                                    .await?
+                                    .expect(&format!(
+                                        "FsSplit with id {} should be in the state table.",
+                                        split_id
+                                    ));
```

Comment:
as the experiment shows, it is likely to hit the branch here. Maybe fill `split_id` in cache as default behavior?